src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,883 @@
+            raise ValueError("Input must be a 5D tensor of shape (batch_size, 2, num_channels, height, width)")
+
+        batch_size, _, channels, height, width = pixel_values.shape
+
+        list_matches_0 = []
+        list_matches_1 = []
+        list_matching_scores_0 = []
+        list_matching_scores_1 = []
+        list_keypoints_0 = []
+        list_keypoints_1 = []
+        list_hidden_states = []
+        list_attentions = []
+
+        for image_pair in pixel_values:
+            keypoint_detection_output = self.keypoint_detector(
+                image_pair,
+                output_hidden_states=output_hidden_states,
+                return_dict=return_dict,
+            )
+
+            keypoints, scores, descriptors, mask = keypoint_detection_output[:4]
+
+            image0_indices = torch.nonzero(mask[0]).squeeze()
+            image0_keypoints = torch.unsqueeze(keypoints[0][image0_indices], dim=0)
+            image0_descriptors = torch.unsqueeze(descriptors[0][image0_indices], dim=0)
+            image0_scores = torch.unsqueeze(scores[0][image0_indices], dim=0)
+            image1_indices = torch.nonzero(mask[1]).squeeze()
+            image1_keypoints = torch.unsqueeze(keypoints[1][image1_indices], dim=0)
+            image1_descriptors = torch.unsqueeze(descriptors[1][image1_indices], dim=0)
+            image1_scores = torch.unsqueeze(scores[1][image1_indices], dim=0)
+
+            match_image_output = self.match_image_pair(
+                image0_keypoints,
+                image0_descriptors,
+                image0_scores,
+                image1_keypoints,
+                image1_descriptors,
+                image1_scores,
+                height,
+                width,
+                output_attentions=output_attentions,
+                output_hidden_states=output_hidden_states,
+            )
+
+            matches_0, matches_1, matching_scores_0, matching_scores_1, hidden_states, attentions = match_image_output
+            list_matches_0.append(matches_0)
+            list_matches_1.append(matches_1)
+            list_matching_scores_0.append(matching_scores_0)
+            list_matching_scores_1.append(matching_scores_1)
+            list_keypoints_0.append(image0_keypoints)
+            list_keypoints_1.append(image1_keypoints)
+            list_hidden_states.append(hidden_states)
+            list_attentions.append(attentions)
+
+        attentions, hidden_states, keypoints, matches, matches_mask, matching_scores = self.batch_output(
+            batch_size,
+            list_attentions,
+            list_hidden_states,
+            list_keypoints_0,
+            list_keypoints_1,
+            list_matches_0,
+            list_matches_1,
+            list_matching_scores_0,
+            list_matching_scores_1,
+            pixel_values,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+        )
```

Comment:
This all feels a bit convoluted - why do we need to take the outputs of `match_image_pair`, create a bunch of lists and the pass all those lists to another method to batch them together. Couldn't we create batched inputs directly in `match_image_output`?

Reply:
Changes have been made to prevent this. Keypoint detection happens for all images beforehand, based on the maximum number of keypoints, all variables are initialized and filled as we match the images